File: core/src/com/unibo/s3/main_system/modules/MenuModule.java
package com.unibo.s3.main_system.modules;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.InputMultiplexer;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.scenes.scene2d.utils.ChangeListener;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.Align;
import com.kotcrab.vis.ui.VisUI;
import com.kotcrab.vis.ui.widget.*;
import com.unibo.s3.Main;

import java.util.HashMap;

public class MenuModule extends BasicModuleWithGui {
    private SpriteBatch textBatch;
    private World world;
    private BitmapFont font;
    private boolean enabled = true;

    private int guardsNum = 5;
    private int thiefsNum = 1;
    private boolean pause = false;
    private boolean simulation = false;
    private boolean mazeMap = true;

    @Override
    public void init(Main owner) {
        super.init(owner);
        this.world = new World(new Vector2(0,0), true);
        font = new BitmapFont();
        font.getRegion().getTexture().setFilter(Texture.TextureFilter.Linear, Texture.TextureFilter.Linear);
        font.getData().setScale(1.5f);
        VisUI.load();
        //initMenuGUI();
        initSettingsGUI();
    }

    private void initSettingsGUI() {
        VisWindow table = new VisWindow("Settings");
        //table.setDebug(true);
        table.setMovable(false);
        Label title = table.getTitleLabel();
        title.setColor(Color.GREEN);
        System.out.println("align = " + title.getLabelAlign());
        title.setAlignment(Align.center);

        table.row();
        VisSlider guardsNumS = new VisSlider(2f, 20f, 1f, false);
        guardsNumS.setValue(this.guardsNum);
        VisLabel labGuardsNum = new VisLabel(" " + String.format("%02d", (int) guardsNumS.getValue()));
        table.add(new VisLabel("Guards number: "));
        table.row();
        VisTable tableGuardum = new VisTable();
        tableGuardum.add(guardsNumS).padLeft(10);
        tableGuardum.add(labGuardsNum).padRight(10).padLeft(5);
        table.add(tableGuardum);
        guardsNumS.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                guardsNum = (int) guardsNumS.getValue();
                labGuardsNum.setText(" "+String.format("%02d", guardsNum));
            }
        });

        table.row();
        VisSlider thiefsNumS = new VisSlider(1f, 20f, 1f, false);
        thiefsNumS.setValue(this.thiefsNum);
        VisLabel labThiefsNum = new VisLabel(" " + String.format("%02d", (int) thiefsNumS.getValue()));
        table.add(new VisLabel("Thiefs number: ")).padTop(10);
        table.row();
        VisTable tableThiefsNum = new VisTable();
        tableThiefsNum.add(thiefsNumS);
        tableThiefsNum.add(labThiefsNum);
        table.add(tableThiefsNum);
        thiefsNumS.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                thiefsNum = (int) thiefsNumS.getValue();
                labThiefsNum.setText(" "+String.format("%02d", thiefsNum));
            }
        });

        table.row();
        /*table.add(new VisLabel("Map tipe: "));
        table.row();*/
        VisTable tableThiefType = new VisTable();
        ButtonGroup<VisRadioButton> group = new ButtonGroup<>();
        VisRadioButton simThief = new VisRadioButton("Simulated");
        group.add(simThief);
        VisRadioButton pilThief = new VisRadioButton("Piloted");
        group.add(pilThief);
        tableThiefType.add(new VisLabel("Thief type: "));
        tableThiefType.add(simThief).padRight(5);
        tableThiefType.add(pilThief).padLeft(5);
        simThief.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if(simThief.isChecked())
                    simulation = true;
                else
                    simulation = false;
            }
        });
        table.add(tableThiefType).padTop(10).padLeft(5).padRight(5);

        //60x60, 80x60
        table.row();
        VisTable tableMapDimension = new VisTable();

        int numDimension = 2;
        String[] dimensionS = new String[numDimension];
        Integer[][] dimensionI = new Integer[numDimension][2];
        dimensionS[0] = "60x60";
        dimensionI[0][0] = 60;
        dimensionI[0][1] = 60;

        dimensionS[1] = "80x60";
        dimensionI[1][0] = 80;
        dimensionI[1][1] = 60;

        VisSelectBox<String> mapDimension = new VisSelectBox<>();
        mapDimension.setItems(dimensionS);
        tableMapDimension.add(new VisLabel("Map dimensions: "));
        tableMapDimension.add(mapDimension);
        table.add(tableMapDimension).padTop(10);
        mapDimension.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                int indexSelected = mapDimension.getSelectedIndex();
                System.out.println(dimensionI[indexSelected][0] + "x" + dimensionI[indexSelected][1]);
            }
        });

        //labirinto: maze - stanze: rooms
        table.row();
        /*table.add(new VisLabel("Map tipe: "));
        table.row();*/
        VisTable tableMapType = new VisTable();
        ButtonGroup<VisRadioButton> groupMaps = new ButtonGroup<>();
        VisRadioButton mazeCheck = new VisRadioButton("Maze");
        groupMaps.add(mazeCheck);
        VisRadioButton roomCheck = new VisRadioButton("Rooms");
        groupMaps.add(roomCheck);
        tableMapType.add(new VisLabel("Map type: "));
        tableMapType.add(mazeCheck).padRight(5);
        tableMapType.add(roomCheck).padLeft(5);
        mazeCheck.addListener(new ChangeListener() {
            @Override
            public void changed(ChangeEvent event, Actor actor) {
                if(mazeCheck.isChecked())
                    mazeMap = true;
                else
                    mazeMap = false;
            }
        });
        table.add(tableMapType).padTop(10);

        table.row();
        VisTextButton okButton = new VisTextButton("Start");
        table.add(okButton).padTop(10).padBottom(10);
        okButton.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                super.clicked(event, x, y);
            }
        });
        table.pack();

        GUI.addActor(table);

        table.centerWindow();
        //table.setPosition((Gdx.graphics.getWidth()/2)-(table.getWidth()/2), (Gdx.graphics.getHeight()/2)-(table.getHeight()/2));
    }

    private void initMenuGUI() {
        VisWindow table = new VisWindow("Menu");
        //table.setDebug(true);
        table.setMovable(false);
        Label title = table.getTitleLabel();
        title.setColor(Color.GREEN);
        title.setAlignment(Align.center);

        table.row();
        String pauseString = "Pause";
        String startString = "Start";
        VisTextButton buttonPause = new VisTextButton(pauseString);
        table.add(buttonPause).padTop(10);
        buttonPause.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                if(pause)
                    buttonPause.setText(pauseString);
                else
                    buttonPause.setText(startString);
                pause = !pause;
            }
        });

        table.row();
        VisTextButton buttonStop = new VisTextButton("STOP");
        table.add(buttonStop).padTop(10);

        table.row();
        VisCheckBox debugView = new VisCheckBox(" View debug");
        table.add(debugView).padTop(10).padBottom(10).padLeft(10).padRight(10);

        GUI.addActor(table);

        table.pack();
        table.setPosition(50, Gdx.graphics.getHeight() - 200);
    }

    @Override
    public void attachInputProcessors(InputMultiplexer inputMultiplexer) {
        super.attachInputProcessors(inputMultiplexer);
        inputMultiplexer.addProcessor(this);
    }

    @Override
    public void enable(boolean enabled) {
        this.enabled = enabled;
    }

    @Override
    public boolean isEnabled() {
        return this.enabled;
    }


    public void setTextRenderer(SpriteBatch b) {
        textBatch = b;
    }
}

File: core/src/com/unibo/s3/main_system/modules/MenuModule.scala
package com.unibo.s3.main_system.modules

import com.badlogic.gdx.{Gdx, InputMultiplexer}
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.scenes.scene2d.{Actor, InputEvent}
import com.badlogic.gdx.scenes.scene2d.ui.{ButtonGroup, Label}
import com.badlogic.gdx.scenes.scene2d.utils.{ChangeListener, ClickListener}
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.Align
import com.kotcrab.vis.ui.util.ToastManager
import com.kotcrab.vis.ui.widget._
import com.kotcrab.vis.ui.widget.toast.Toast
import com.unibo.s3.Main
import com.unibo.s3.testbed.ui.{AdaptiveSizeActor, Anchorable, TopLeft}
import com.unibo.s3.testbed.ui.KeyHelpTable

sealed trait MenuEvent
case class Start(guardsNum: Int, thiefsNum: Int, simulation: Boolean, mapDimension: Vector2, mazeTypeMap: Boolean) extends MenuEvent
case class Pause(pause: Boolean) extends MenuEvent
case class ViewDebug(debug: Boolean) extends MenuEvent

class MenuModule(listener: MenuEvent => Unit) extends BasicModuleWithGui{
  private var enabled = true
  private var windowSettings, windowMenu: VisWindow = _
  private var menuWindowPane: AdaptiveSizeActor with Anchorable = _
  private var guardsNum = 5
  private var thiefsNum = 1
  private var simulation = true
  private var dimensionMap: Vector2 = _
  private var mazeMap = true

  private var pause = false
  private var viewDebug = false

  override def init(owner: Main): Unit = {
    super.init(owner)
    //this.world = new World(new Vector2(0,0), true);
    initMenuGUI()
    initSettingsGUI()
  }

  private def initSettingsGUI() = {
    windowSettings = new VisWindow("Settings")
    //windowSettings.setDebug(true);
    windowSettings.setMovable(false)
    val title = windowSettings.getTitleLabel
    title.setColor(Color.GREEN)
    title.setAlignment(Align.center)

    //windowSettings.row
    val guardsNumS = new VisSlider(2f, 20f, 1f, false)
    guardsNumS.setValue(this.guardsNum)
    val labGuardsNum = new VisLabel(" " + "%02d".format(guardsNum))
    windowSettings.add(new VisLabel("Guards number: ")).row()
    val tableGuardum = new VisTable
    tableGuardum.add(guardsNumS).padLeft(10)
    tableGuardum.add(labGuardsNum).padRight(10).padLeft(5)
    windowSettings.add(tableGuardum).row()
    guardsNumS.addListener(new ChangeListener() {
      override def changed(event: ChangeListener.ChangeEvent, actor: Actor): Unit = {
        guardsNum = guardsNumS.getValue.toInt
        labGuardsNum.setText(" " + "%02d".format(guardsNum))
      }
    })

    val thiefsNumS = new VisSlider(1f, 20f, 1f, false)
    thiefsNumS.setValue(this.thiefsNum)
    val labThiefsNum = new VisLabel(" " + "%02d".format(thiefsNum))
    windowSettings.add(new VisLabel("Thiefs number: ")).padTop(10).row()
    val tableThiefsNum = new VisTable
    tableThiefsNum.add(thiefsNumS)
    tableThiefsNum.add(labThiefsNum)
    windowSettings.add(tableThiefsNum).row()
    thiefsNumS.addListener(new ChangeListener() {
      override def changed(event: ChangeListener.ChangeEvent, actor: Actor): Unit = {
        thiefsNum = thiefsNumS.getValue.toInt
        labThiefsNum.setText(" " + "%02d".format(thiefsNum))
      }
    })

    val tableThiefType = new VisTable
    val group = new ButtonGroup[VisRadioButton]
    val simThiefRB = new VisRadioButton("Simulated")
    group.add(simThiefRB)
    val pilThiefRB = new VisRadioButton("Piloted")
    group.add(pilThiefRB)
    tableThiefType.add(new VisLabel("Thief type: "))
    tableThiefType.add(simThiefRB).padRight(5)
    tableThiefType.add(pilThiefRB).padLeft(5)
    windowSettings.add(tableThiefType).padTop(10).padLeft(5).padRight(5).row()
    simThiefRB.addListener(new ChangeListener() {
      override def changed(event: ChangeListener.ChangeEvent, actor: Actor): Unit = {
        simulation = simThiefRB.isChecked
      }
    })

    //60x60, 80x60
    val tableMapDimension = new VisTable
    val numDimension = 2
    val dimensionS = new Array[String](numDimension)
    val dimensionI = new Array[Vector2](numDimension)
    var index = 0
    dimensionI.add(new Vector2(60,60))
    dimensionS.add(dimensionI.get(index).x.toInt + "x" + dimensionI.get(index).y.toInt)
    dimensionMap = dimensionI.get(index)
    index += 1
    dimensionI.add(new Vector2(80,60))
    dimensionS.add(dimensionI.get(index).x.toInt + "x" + dimensionI.get(index).y.toInt)
    val mapDimensionSB = new VisSelectBox[String]
    mapDimensionSB.setItems(dimensionS)
    tableMapDimension.add(new VisLabel("Map dimensions: "))
    tableMapDimension.add(mapDimensionSB)
    windowSettings.add(tableMapDimension).padTop(10).row()
    mapDimensionSB.addListener(new ChangeListener() {
      override def changed(event: ChangeListener.ChangeEvent, actor: Actor): Unit = {
        val indexSelected = mapDimensionSB.getSelectedIndex
        //System.out.println(dimensionI(indexSelected)(0) + "x" + dimensionI(indexSelected)(1))
        dimensionMap = dimensionI.get(indexSelected)
      }
    })

    //labirinto: maze - stanze: rooms
    val tableMapType = new VisTable
    val groupMaps = new ButtonGroup[VisRadioButton]
    val mazeCheck = new VisRadioButton("Maze")
    groupMaps.add(mazeCheck)
    val roomCheck = new VisRadioButton("Rooms")
    groupMaps.add(roomCheck)
    tableMapType.add(new VisLabel("Map type: "))
    tableMapType.add(mazeCheck).padRight(5)
    tableMapType.add(roomCheck).padLeft(5)
    windowSettings.add(tableMapType).padTop(10).row()
    mazeCheck.addListener(new ChangeListener() {
      override def changed(event: ChangeListener.ChangeEvent, actor: Actor): Unit = {
        mazeMap = mazeCheck.isChecked
      }
    })

    val okButton = new VisTextButton("Start!", "blue")
    windowSettings.add(okButton).padTop(10).padBottom(10).row()
    okButton.addListener(new ClickListener() {
      override def clicked(event: InputEvent, x: Float, y: Float): Unit = {
        listener(Start(guardsNum, thiefsNum, simulation, dimensionMap, mazeMap))
        windowSettings.fadeOut(1f)
        windowMenu.setVisible(true)
        windowMenu.fadeIn(1f)
      }
    })
    windowSettings.pack()
    gui.addActor(windowSettings)
    windowSettings.centerWindow
    //windowSettings.setPosition((Gdx.graphics.getWidth()/2)-(windowSettings.getWidth()/2), (Gdx.graphics.getHeight()/2)-(windowSettings.getHeight()/2));
  }

  private def initMenuGUI() = {
    windowMenu = new VisWindow("Menu")
    windowMenu.setMovable(false)
    windowMenu.setVisible(false)
    val title = windowMenu.getTitleLabel
    title.setColor(Color.GREEN)
    title.setAlignment(Align.center)

    val pauseString = "Pause"
    val startString = "Resume"
    val buttonPause = new VisTextButton(pauseString)
    windowMenu.add(buttonPause).padTop(10).row()
    buttonPause.addListener(new ClickListener() {
      override def clicked(event: InputEvent, x: Float, y: Float): Unit = {
        if (pause) buttonPause.setText(pauseString)
        else buttonPause.setText(startString)
        pause = !pause
        listener(Pause(pause))
      }
    })

    val debugView = new VisCheckBox(" View debug")
    windowMenu.add(debugView).padTop(10).padLeft(10).padRight(10).row()
    debugView.addListener(new ChangeListener {
      override def changed(event: ChangeListener.ChangeEvent, actor: Actor): Unit = {
        if(debugView.isChecked) viewDebug = true
        else viewDebug = false
        listener(ViewDebug(viewDebug))
      }
    })

    val shortButton = new VisTextButton("Shortcut")
    windowMenu.add(shortButton).padTop(10).padBottom(10)
    shortButton.addListener(new ClickListener() {

      def centerToast(toast: Toast, toastManager: ToastManager): Unit = {
        toast.getMainTable.setY(
          gui.getHeight - toast.getMainTable.getPrefHeight - toastManager.getScreenPadding)
        toast.getMainTable.setX((gui.getWidth / 2) - toast.getMainTable.getPrefWidth / 2 )
      }

      def openShortcut(): Unit = {
        val toastManager = new ToastManager(gui)
        toastManager.setAlignment(Align.center)

        val keys = new KeyHelpTable(true)
        keys.addKeyBinding("p", "pause")
        keys.addKeyBinding("q", "zoom in")
        keys.addKeyBinding("r", "zoom out")
        keys.addKeyBinding("arrow-left", "move camera left")
        keys.addKeyBinding("arrow-right", "move camera right")
        keys.addKeyBinding("arrow-up", "move camera up")
        keys.addKeyBinding("arrow-down", "move camera down")
        keys.addKeyBinding("mouse-wheel-move", "brightness control")
        keys.addKeyBinding("mouse-left", "spawn guard (only debug view)")
        keys.addKeyBinding("t", "spawn thief (only debug view)")
        keys.align(Align.center)

        val t = new Toast("dark", keys)
        toastManager.show(t)
        centerToast(t, toastManager)
      }

      override def clicked(event: InputEvent, x: Float, y: Float): Unit = {
        openShortcut()
      }
    })

    gui.addActor(windowMenu)
    windowMenu.pack()
    //windowMenu.setPosition(50, Gdx.graphics.getHeight - 200)

    menuWindowPane = new AdaptiveSizeActor(windowMenu) with Anchorable
    menuWindowPane.setSize(10, 20) //10 % 20 % of stage size
    menuWindowPane.setPadding(10,0)
    menuWindowPane.setAnchor(TopLeft)
  }

  override def resize(newWidth: Int, newHeight: Int): Unit = {
    super.resize(newWidth, newHeight)
    menuWindowPane.resize(newWidth, newHeight)
  }

  override def attachInputProcessors(inputMultiplexer: InputMultiplexer): Unit = {
    super.attachInputProcessors(inputMultiplexer)
    inputMultiplexer.addProcessor(this)
  }

  override def enable(enabled: Boolean): Unit = {
    this.enabled = enabled
  }

  override def isEnabled: Boolean = this.enabled
}
